qa/tasks/vstart_runner.py

```diff
@@ -176,7 +176,10 @@ def wait(self):
                 return
 
         out, err = self.subproc.communicate()
-        out, err = out.decode(), err.decode()
```

Comment:
i'd suggest 
```py
out = out.decode(errors='ignore')
err = err.decode(errors='ignore')
```

i guess the reason why decode failed in https://tracker.ceph.com/issues/43567 is that cephfs-shell added some ANSI escape codes which is not in the range of ASCII.

and your change only works for python2 where `Popen.communicate()` returns `str`s. while in python3, it returns bytes. and if we don't decode property, it still fails with the same error. so we do need to ignore these escape codes here.

Reply:
The suggestions looks better. But isn't there a chance that a valid error might be ignored by passing `errors=ignore`.

Reply:
> But isn't there a chance that a valid error might be ignored by passing `errors=ignore`.

another option is to filter out the escape chars from the output from cephfs-shell, and keep the "strict" default.

Reply:
I'll try that instead.

Reply:
There's a better fix. Working on that now.

Reply:
There's a better fix. Working on that now. It is to do `self.mount_a.run_shell([cmdargs])` than `self.get_cephfs_shell_output('ls')`.

Reply:
>  guess the reason why decode failed in https://tracker.ceph.com/issues/43567 is that cephfs-shell added some ANSI escape codes which is not in the range of ASCII.

Contents of stderr -

`'mkdir: cannot create directory \xe2\x80\x98/tmp/cephfstest2xG8fn\xe2\x80\x99: File exists\n'`

The hexcode(?) represents `‘` and `’`.

Reply:
I've refactored the test but looks like we'll need to fix it in case of other command failures -

```
(Pdb) p err
"dd: unrecognized operand \xe2\x80\x981M\xe2\x80\x99\nTry 'dd --help' for more information.\n"
```

Saw the above case while testing my patch. Perhaps the fix is to recognize some of these hexcodes and do a replace instead. We could replace above hexcodes by `'` and lookout for more cases as we find more errors. @batrick @tchaikov Any opinion?

Reply:
After [the second commit](https://github.com/ceph/ceph/pull/32612/commits/2141d4d628d10a4f394024bd39a388dab1513efa), the exception doesn't occur anymore -

```
2020-01-14 20:39:40,032.032 INFO:__main__:CommandFailedError: Command failed with status 1: ['dd', 'if=/dev/urandom', 'of=/tmp/cephfstestCmLsI0/dump1', '1M', 'count=1']
```

Reply:
@rishabh-d-dave what's the output of `locale`? coreutils uses gettext to do the i10n. and the error messages are displayed according to your locale settings, see https://docs.python.org/3/library/gettext.html#gnu-gettext-api . you could use `locale.getlocale()` to get it programmatically . see https://docs.python.org/3/library/locale.html#locale.getlocale .

i don't think we should customize the filter to adapt to the locales. it's just magic numbers in the code. instead i think we should set it to "C" before using `subproc.communicate()`
```py
locale.setlocale(locale.LC_ALL, 'C')
```
if we care about the output and try to collect it using `StringIO`.

Reply:
with filter, without locale
---------------------------
code
```
        out, err = self.subproc.communicate()
        out, err = rm_nonascii_chars(out), rm_nonascii_chars(err)
        out, err = out.decode(), err.decode()
``` 

output
```
    2020-01-17 14:36:54,988.988 INFO:__main__:======================================================================
    2020-01-17 14:36:54,988.988 INFO:__main__:ERROR: test_put_and_get_without_target_directory (tasks.cephfs.test_cephfs_shell.TestGetAndPut)
    2020-01-17 14:36:54,988.988 INFO:__main__:----------------------------------------------------------------------
    2020-01-17 14:36:54,988.988 INFO:__main__:Traceback (most recent call last):
    2020-01-17 14:36:54,988.988 INFO:__main__:  File "/home/rishabh/repos/ceph/fs-shell-retval/qa/tasks/cephfs/test_cephfs_shell.py", line 186, in test_put_and_get_without_target_directory
    2020-01-17 14:36:54,988.988 INFO:__main__:    self.mount_a.run_shell(['mkdir', 'a'])
    2020-01-17 14:36:54,988.988 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 717, in run_shell
    2020-01-17 14:36:54,988.988 INFO:__main__:    omit_sudo=omit_sudo)
    2020-01-17 14:36:54,988.988 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 396, in run
    2020-01-17 14:36:54,988.988 INFO:__main__:    proc.wait()
    2020-01-17 14:36:54,988.988 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 196, in wait
    2020-01-17 14:36:54,988.988 INFO:__main__:    raise CommandFailedError(self.args, self.exitstatus)
    2020-01-17 14:36:54,988.988 INFO:__main__:CommandFailedError: Command failed with status 1: ['mkdir', 'a']
 ```

without filter, with locale
---------------------------
code
```
        import locale
        locale.setlocale(locale.LC_ALL, 'C')
        out, err = self.subproc.communicate()
        out, err = out.decode(), err.decode()
 ```

output
```
    2020-01-17 14:48:30,838.838 INFO:__main__:======================================================================
    2020-01-17 14:48:30,838.838 INFO:__main__:ERROR: test_put_and_get_without_target_directory (tasks.cephfs.test_cephfs_shell.TestGetAndPut)
    2020-01-17 14:48:30,838.838 INFO:__main__:----------------------------------------------------------------------
    2020-01-17 14:48:30,838.838 INFO:__main__:Traceback (most recent call last):
    2020-01-17 14:48:30,838.838 INFO:__main__:  File "/home/rishabh/repos/ceph/fs-shell-retval/qa/tasks/cephfs/test_cephfs_shell.py", line 186, in test_put_and_get_without_target_directory
    2020-01-17 14:48:30,838.838 INFO:__main__:    self.mount_a.run_shell(['mkdir', 'a'])
    2020-01-17 14:48:30,838.838 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 718, in run_shell
    2020-01-17 14:48:30,838.838 INFO:__main__:    omit_sudo=omit_sudo)
    2020-01-17 14:48:30,839.839 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 397, in run
    2020-01-17 14:48:30,839.839 INFO:__main__:    proc.wait()
    2020-01-17 14:48:30,839.839 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 186, in wait
    2020-01-17 14:48:30,839.839 INFO:__main__:    out, err = out.decode(), err.decode()
    2020-01-17 14:48:30,839.839 INFO:__main__:UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 31: ordinal not in range(128)
```

Reply:
@rishabh-d-dave you need to set the locale before `subprocess.Popen()`.

Reply:
Got same result -

```
diff --git a/qa/tasks/vstart_runner.py b/qa/tasks/vstart_runner.py
index dfc01a0615..ecb13de1fb 100644
--- a/qa/tasks/vstart_runner.py
+++ b/qa/tasks/vstart_runner.py
@@ -181,7 +181,7 @@ class LocalRemoteProcess(object):
                 return
 
         out, err = self.subproc.communicate()
-        out, err = rm_nonascii_chars(out), rm_nonascii_chars(err)
+        #out, err = rm_nonascii_chars(out), rm_nonascii_chars(err)
         out, err = out.decode(), err.decode()
         self.stdout.write(out)
         self.stderr.write(err)
@@ -357,6 +357,9 @@ class LocalRemote(object):
 
         log.info("Running {0}".format(args))
 
+        import locale
+        locale.setlocale(locale.LC_ALL, 'C')
+
         if shell:
             subproc = subprocess.Popen(quote(args),
                                        stdout=subprocess.PIPE,
```

```
2020-01-17 17:54:04,927.927 INFO:__main__:======================================================================
2020-01-17 17:54:04,927.927 INFO:__main__:ERROR: test_put_and_get_without_target_directory (tasks.cephfs.test_cephfs_shell.TestGetAndPut)
2020-01-17 17:54:04,927.927 INFO:__main__:----------------------------------------------------------------------
2020-01-17 17:54:04,928.928 INFO:__main__:Traceback (most recent call last):
2020-01-17 17:54:04,928.928 INFO:__main__:  File "/home/rishabh/repos/ceph/fs-shell-retval/qa/tasks/cephfs/test_cephfs_shell.py", line 186, in test_put_and_get_without_target_directory
2020-01-17 17:54:04,928.928 INFO:__main__:    self.mount_a.run_shell(['mkdir', 'a'])
2020-01-17 17:54:04,928.928 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 720, in run_shell
2020-01-17 17:54:04,928.928 INFO:__main__:    omit_sudo=omit_sudo)
2020-01-17 17:54:04,928.928 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 399, in run
2020-01-17 17:54:04,928.928 INFO:__main__:    proc.wait()
2020-01-17 17:54:04,928.928 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 185, in wait
2020-01-17 17:54:04,928.928 INFO:__main__:    out, err = out.decode(), err.decode()
2020-01-17 17:54:04,928.928 INFO:__main__:UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 31: ordinal not in range(128)
```